crates/sound/src/device.rs

```diff
@@ -70,49 +76,301 @@ impl VhostUserSoundThread {
+    fn process_control(
+        &self,
+        vring: &VringRwLock,
+        audio_backend: &RwLock<Box<dyn AudioBackend + Send + Sync>>,
+    ) -> IoResult<bool> {
+        let requests: Vec<SoundDescriptorChain> = vring
             .get_mut()
             .get_queue_mut()
             .iter(self.mem.as_ref().unwrap().memory())
             .map_err(|_| Error::DescriptorNotFound)?
             .collect();
-        dbg!(&requests);
 
-        Ok(true)
+        if requests.is_empty() {
+            return Ok(true);
+        }
+
+        // Reply to some requests right away, and defer others to the audio backend (for
+        // example PcmRelease needs to complete all I/O before replying.
+        //
+        // Mark `any` as true if we need to reply to any request right away so we can
+        // signal the queue as used.
+        let mut any = false;
+
+        for desc_chain in requests {
+            let descriptors: Vec<_> = desc_chain.clone().collect();
+            // Request descriptor.
+            let desc_request = descriptors[0];
+            if desc_request.is_write_only() {
+                return Err(Error::UnexpectedWriteOnlyDescriptor(0).into());
+            }
+
+            let request = desc_chain
+                .memory()
+                .read_obj::<VirtioSoundHeader>(desc_request.addr())
+                .map_err(|_| Error::DescriptorReadFailed)?;
+
+            // Keep track of bytes that will be written in the VQ.
+            let mut len = 0;
+
+            // Reply header descriptor.
+            let desc_hdr = descriptors[1];
+            if !desc_hdr.is_write_only() {
+                return Err(Error::UnexpectedReadableDescriptor(1).into());
+            }
+            len += desc_hdr.len();
+
+            let mut resp = VirtioSoundHeader {
+                code: VIRTIO_SND_S_OK.into(),
+            };
+
+            let code = ControlMessageKind::try_from(request.code).map_err(Error::from)?;
+            match code {
+                ControlMessageKind::JackInfo => {
+                    resp.code = VIRTIO_SND_S_NOT_SUPP.into();
+                }
+                ControlMessageKind::JackRemap => {
+                    resp.code = VIRTIO_SND_S_NOT_SUPP.into();
+                }
+                ControlMessageKind::PcmInfo => {
+                    let request = desc_chain
+                        .memory()
+                        .read_obj::<VirtioSoundQueryInfo>(desc_request.addr())
+                        .map_err(|_| Error::DescriptorReadFailed)?;
+
+                    let start_id = u32::from(request.start_id) as usize;
+                    let count = u32::from(request.count) as usize;
+                    let streams = self.streams.read().unwrap();
+                    if streams.len() <= start_id || streams.len() <= start_id + count {
+                        resp.code = VIRTIO_SND_S_BAD_MSG.into();
+                    } else {
+                        let desc_response = descriptors[2];
```

Comment:
Are we sure that we always have this layout?

I mean: descriptor[1] header and descriptor[2] payload.

Or is it an implementation detail of the linux driver?

I ask because in virtio-vsock we had a similar problem: the specification doesn't define these details and the driver initially used 2 separate descriptors for header and payload, then it started to use only one descriptor and vhost-user-vsock didn't work.

That's fine for now, but I would put a comment if that's the case.

Reply:
I think so, at least from my understanding of the spec wording (I haven't worked with virtqueues directly before, caveat emptor)

```
A generic control message consists of a request part and a response part.

A request part has, or consists of, a common header containing the following device-readable field:

code
    specifies a device request type (VIRTIO_SND_R_*).

A response part has, or consists of, a common header containing the following device-writable field:

code
    indicates a device request status (VIRTIO_SND_S_*).

The status field can take one of the following values:

    VIRTIO_SND_S_OK - success.
    VIRTIO_SND_S_BAD_MSG - a control message is malformed or contains invalid parameters.
    VIRTIO_SND_S_NOT_SUPP - requested operation or parameters are not supported.
    VIRTIO_SND_S_IO_ERR - an I/O error occurred.

The request part may be followed by an additional device-readable payload, and the response part may be followed by an additional device-writable payload. 
```

So in this case the generic ctrl header is followed by the additional device-readable payload, and then the response comes after.

Reply:
Okay, so even in this case it's an implementation detail of the driver. Nothing prevents it from using a single writable descriptor to contain both the header and the response payload, right?

Reply:
Almost all rust-vmm daemons get this wrong in one way or another... We would really need something like https://github.com/rust-vmm/vm-virtio/pull/33 to prevent daemons from needing to worry about this stuff...

Reply:
Yep, I came to the same conclusion.

So that's okay in my opinion for now, but we should put comments where we make this assumption to say that's how the driver works in Linux 6.4.

WDYT?

Reply:
> So that's okay in my opinion for now, but we should put comments where we make this assumption to say that's how the driver works in Linux 6.4.

Sounds fine to me at least. Everyone else seems to do the same, just without the comment :).

Reply:
> Nothing prevents it from using a single writable descriptor to contain both the header and the response payload, right?

It does say each "element" in the virtqueue is placed on a separate descriptor but could definitely be more specific:

```
> 2.7.13.1 Placing Buffers Into The Descriptor Table

A buffer consists of zero or more device-readable physically-contiguous elements followed by zero or more physically-contiguous device-writable elements (each has at least one element). This algorithm maps it into the descriptor table to form a descriptor chain:

for each buffer element, b:

    Get the next free descriptor table entry, d
    Set d.addr to the physical address of the start of b
    Set d.len to the length of b.
    If b is device-writable, set d.flags to VIRTQ_DESC_F_WRITE, otherwise 0.
    If there is a buffer element after this:
        Set d.next to the index of the next free descriptor element.
        Set the VIRTQ_DESC_F_NEXT bit in d.flags.

In practice, d.next is usually used to chain free descriptors, and a separate count kept to check there are enough free descriptors before beginning the mappings.
```

Reply:
I think this means it's not implementation specific

Reply:
I digged this out of my IRC history:
```
Ablu: The "Supplying Buffers to The Device" subsections for both splitted and packaged virtqueues seems to suggest that a single buffer can be split across multiple descriptors... 
vireshk: Do you have a reference to something that explains the 1 buffer == 1 descriptor? I guess naturally, most of the time that will be the case, just trying to read what the spec says literally...

Ablu: I think I found it... https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-280004 says that it can be split up arbitrarily. So I think the GPIO device, should (in theory) also be able to handle >2 descriptors I think? Though then... https://github.com/rust-vmm/vm-virtio/blob/main/crates/devices/virtio-blk/src/request.rs#L236-L244 also looks "wrong" since nothing specifies that the status has to be in it's own descriptor? So chances are that I am still missing something...

Ablu: I asked on rust-vmm slack about the virtio-blk code... 

Ablu: https://github.com/rust-vmm/vm-virtio/issues/216 is the answer :). So I am tempted to claim that the GPIO code asserting len == 2 is "wrong".
```
Unfortunately the Slack thread about this got lost since it is older than 90 days. But maybe this gives some context :).

Reply:
@epilys header and payload could be on a single buffer, though, couldn't they?

The important thing is not to mix writable and readable buffers, but then for virtqueue it can contain anything as long as it's contiguous memory. So nothing should prevent the driver from using a single contiguous memory area and expecting the device to write both the header and the response payload there.

Reply:
are they considered the same element? (response header + payload)

Reply:
Yes, I think what you quoted above talks about an "element" in the virtqueue. So the element there is a descriptor chain. The chain then can have any form (as long as r/w constraints are satisfied) and consist of multiple descriptors.

Reply:
an element for the virtqueue are just contiguous bytes, then as each device uses them, they are a higher level of abstraction

Reply:
I see, thanks!